File: 0070-climbing-stairs/0070-climbing-stairs.java
class Solution {
    
    // create a hashmap to store the previously calculated number of steps
    HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
    
    public int climbStairs(int n) {
        // if number of steps already belong in the map then return it
        if (map.containsKey(n)) {
            return map.get(n);
        }
        // base cases of 0,1,2
        if (n == 0) {
            return 0;
        }
        if (n == 1) {
            return 1;
        }
        if (n == 2) {
            return 2;
        }
        // else we calculate the number of steps (which is a Fibonacci) and put into the map, then return it 
        else {
            int steps = climbStairs(n - 1) + climbStairs(n - 2);
            map.put(n, steps);
            return steps;
        }
    }
}

File: 0070-climbing-stairs/0070-climbing-stairs.py
class Solution:
    # top down way
    # map = {}
    # def climbStairs(self, n: int) -> int:
    #     if n in self.map:
    #         return self.map[n]
    #     if n == 0:
    #         return 0
    #     if n == 1:
    #         return 1
    #     if n == 2:
    #         return 2
    #     steps = self.climbStairs(n - 1) + self.climbStairs(n - 2)
    #     self.map[n] = steps
    #     return steps

    # bottom down way
    def climbStairs(self, n:int) -> int:
        dp = [1,1]
        for i in range(2,n+1):
            dp.append(dp[i-1] + dp[i-2])
        return dp[n]